futures-02.md

```diff
@@ -0,0 +1,396 @@
+explain some of the desired changes to the `futures` create. Please leave
+comments about specific issues on [the `futures` repo](https://github.com/alexcrichton/futures-rs),
+as detailed discussion could easily overwhelm the RFC thread.
+
+## Explicit Task Context and Wakeup Handles
+
+The `Future`, `Stream`, `AsyncRead`, and `AsyncWrite` traits have been modified
+to take explicit `task::Context` arguments. `task::Context` is a type which
+provides access to
+[task-locals](https://docs.rs/futures/*/futures/macro.task_local.html) and a `Waker`:
+
+```rust
+mod task {
+
+  /// The context for the currently running task.
+  pub struct Context {
+
+      /// Returns a `Waker` which can be used to wake up the current task.
+      pub fn waker(&self) -> Waker { ... }
+
+      /// Runs the provided closure with the task-local pointed to by the
+      /// provided `key`.
+      ///
+      /// `LocalKey`s are acquired using the `task_local` macro.
+      pub fn with_local<T, R, F>(&mut self, key: LocalKey<T>, f: F) -> R
+          where F: FnOnce(&mut T) -> R
+      {
+          ...
+      }
+  }
+
+  /// A handle used to awaken a task.
+  pub struct Waker {
+
+      /// Wakes up the associated task.
+      ///
+      /// This signals that the futures associated with the task are ready to
+      /// be `poll`ed again.
+      pub fn wake(&self) { ... }
+  }
+
+  ...
+}
+```
+
+The modified traits will look as follows:
+
+```rust
+pub trait Future {
+    type Item;
+    type Error;
+    fn poll(&mut self, ctx: &mut Context) -> Poll<Self::Item, Self::Error>;
+}
+
+pub trait Stream {
+    type Item;
+    type Error;
+    fn poll_next(&mut self, ctx: &mut Context) -> Poll<Option<Self::Item>, Self::Error>;
+}
+
+pub trait AsyncRead {
+    type Error;
+    unsafe fn initializer(&self) -> Initializer { ... }
+
+    fn poll_read(&mut self, buf: &mut [u8], ctx: &mut Context)
+        -> Poll<usize, Self::Error>;
+
+    fn poll_vectored_read(&mut self, vec: &mut [&mut IoVec], ctx: &mut Context)
+        -> Poll<usize, Self::Error> { ... }
+}
+
+pub trait AsyncWrite {
+    type Error;
+    fn poll_write(&mut self, buf: &[u8], ctx: &mut Context)
+        -> Poll<usize, Self::Error>;
+
+    fn poll_vectored_write(&mut self, vec: &[&IoVec], ctx: &mut Context)
+        -> Poll<usize, Self::Error> { ... }
+
+    fn poll_close(&mut self, ctx: &mut Context) -> Poll<(), Self::Error>;
+}
+```
+
+In terms of the old `futures` task model, the `Waker` is analogous to the `Task`
+returned by [`task::current`][current], and the `wake` method is analogous to
+the old [`Task::notify`][notify].
+
+Astute readers will notice other changes in the traits shown above.
+These will be addressed in later sections.
+
+[current]: https://docs.rs/futures/0.1.16/futures/task/fn.current.html
+[notify]: https://docs.rs/futures/0.1.16/futures/task/struct.Task.html#method.notify
+
+## Crate Separation
+
+The `futures` crate will be split into several different libraries.
```

Comment:
Could you explicitly say which of these crates are expected to be usable with `#[no_std]`, and which ones will require the standard library?

Reply:
All of them are compatible with `no_std` except for ``futures-channels`.